File: 백준/Gold/14502. 연구소/연구소.py
import sys
from collections import deque

N, M = map(int, sys.stdin.readline().split())

graph = [list(map(int, sys.stdin.readline().split())) for _ in range(N)]

def bfs(cur_graph, n, m):
    bfs_cur_graph = [row[:] for row in cur_graph]

    for cur_n in range(n):
        for cur_m in range(m):
            if bfs_cur_graph[cur_n][cur_m] == 2:
                need_visit = deque([[cur_n, cur_m]])
                while need_visit:
                    visited_n, visited_m = need_visit.popleft()

                    for next_n, next_m in [[visited_n - 1, visited_m], [visited_n + 1, visited_m], [visited_n, visited_m - 1], [visited_n, visited_m + 1]]:
                        if 0 <= next_n < n and 0 <= next_m < m and bfs_cur_graph[next_n][next_m] == 0:
                            bfs_cur_graph[next_n][next_m] = 2
                            need_visit.append([next_n, next_m])

    return sum(row.count(0) for row in bfs_cur_graph)


def back_tracking(cur_graph, cur_n, cur_m, cur_wall, n, m):

    if cur_wall == 3:
        return bfs(cur_graph, n, m)

    result = 0

    for next_n in range(cur_n, n):
        for next_m in range(cur_m if next_n == cur_n else 0, m):
            if cur_graph[next_n][next_m] == 0:
                cur_graph[next_n][next_m] = 1
                result = max(result, back_tracking(cur_graph, next_n, next_m, cur_wall + 1, n, m))
                cur_graph[next_n][next_m] = 0
    return result

print(back_tracking(graph, 0, 0, 0, N, M))

File: 백준/Gold/14502. 연구소/연구소.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int dx[] = {-1, 1, 0, 0};
int dy[] = {0, 0, -1, 1};

int bfs(vector<vector<int>> board, int N, int M){

    vector<vector<bool>> visited(N, vector<bool>(M, false));

    for(int i = 0; i < N; i++){
        for(int j = 0; j < M; j++){
            if(visited[i][j] || board[i][j] != 2)
                continue;

            vector<pair<int, int>> need_visit;
            need_visit.emplace_back(i, j);

            while (!need_visit.empty()){
                pair<int, int> cur_pos = need_visit.back();
                need_visit.pop_back();

                for(int d = 0; d < 4; d++){
                    int next_n = cur_pos.first + dx[d], next_m = cur_pos.second + dy[d];
                    if(0 <= next_n && next_n < N && 0 <= next_m && next_m < M && !visited[next_n][next_m] && board[next_n][next_m] == 0){
                        board[next_n][next_m] = 2;
                        visited[next_n][next_m] = true;
                        need_visit.emplace_back(next_n, next_m);
                    }
                }
            }
        }
    }

    int result = 0;

    for(int i = 0; i < N; i++){
        for(int j = 0; j < M; j++)
            if(board[i][j] == 0) result++;
    }

    return result;
}

int back_tracking(vector<vector<int>> &board, int construct_wall, int cur_n, int cur_m, int N, int M){
    if(construct_wall == 0){
        return bfs(board, N, M);
    }

    int result = 0;

    for(int i = cur_n; i < N; i++){
        for(int j = (i == cur_n) ? cur_m : 0; j < M; j++){
            if(board[i][j] == 0) {
                board[i][j] = 1;
                result = max(result, back_tracking(board, construct_wall - 1, i, j, N, M));
                board[i][j] = 0;
            }
        }
    }
    return result;
}

int main(){
    int N, M;
    cin >> N >> M;

    vector<vector<int>> board(N, vector<int>(M, 0));

    for(int i = 0; i < N; i++){
        for(int j = 0; j < M; j++){
            cin >> board[i][j];
        }
    }

    cout << back_tracking(board, 3, 0, 0, N, M) << endl;
}